contracts/core/src/solve/Inbox.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity =0.8.24;
+
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+import { SafeERC20, IERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { Solve } from "./Solve.sol";
+
+/**
+ * @title Inbox
+ * @notice Entrypoint and alt-mempoool for user solve requests.
+ */
+contract Inbox is ReentrancyGuardUpgradeable {
+    using SafeERC20 for IERC20;
+
+    error NoDeposits();
+    error InvalidCall();
+    error InvalidDeposit();
+    error ZeroDeposit();
+
+    /**
+     * @notice Emitted when a request is created.
+     * @param id        ID of the request.
+     * @param from      Address of the user who created the request.
+     * @param call      Details of the call to be executed on another chain.
+     * @param deposits  Array of deposits backing the request.
+     */
+    event Requested(bytes32 indexed id, address indexed from, Solve.Call call, Solve.Deposit[] deposits);
+
+    /**
+     * @dev uint repr of last assigned request ID.
+     */
+    uint256 internal _lastId;
+
+    /**
+     * @notice Map ID to request.
+     */
+    mapping(bytes32 id => Solve.Request) internal _requests;
+
+    function initialize() public initializer {
+        __ReentrancyGuard_init();
+    }
+
+    /**
+     * @notice Returns the request with the given ID.
+     */
+    function getRequest(bytes32 id) external view returns (Solve.Request memory) {
+        return _requests[id];
+    }
+
+    /**
+     * @notice Open a request to execute a call on another chain, backed by deposits.
+     *  Token deposits are transferred from msg.sender to this inbox.
+     * @param call      Details of the call to be executed on another chain.
+     * @param deposits  Array of deposits backing the request.
+     */
+    function request(Solve.Call calldata call, Solve.TokenDeposit[] calldata deposits)
+        external
+        payable
+        nonReentrant
+        returns (bytes32 id)
+    {
+        if (call.target == address(0)) revert InvalidCall();
+        if (call.destChainId == 0) revert InvalidCall();
+        if (call.data.length == 0) revert InvalidCall();
+        if (deposits.length == 0 && msg.value == 0) revert NoDeposits();
+
+        Solve.Request storage req = _openRequest(msg.sender, call, deposits);
```

Comment:
If `_openRequest` sets storage within its call (referenced [here](https://github.com/omni-network/omni/pull/2348/files#r1826000433)), this should be changed to a memory variable. Using storage here when storage is not being updated just leads to higher read costs.

Reply:
it's a question of when do we resolve the actualy `req.deposits`. Are we resolving it in memory then looping to push . or are we just resolving it once in storage and returning it